packages/walletconnect/src/index.ts

```diff
@@ -73,47 +67,57 @@ export class WalletConnect extends Connector {
     if (error) this.onError?.(error)
   }
 
-  private chainChangedListener = (chainId: number | string): void => {
-    this.actions.update({ chainId: parseChainId(chainId) })
+  private chainChangedListener = (chainId: number): void => {
+    this.actions.update({ chainId })
   }
 
   private accountsChangedListener = (accounts: string[]): void => {
     this.actions.update({ accounts })
   }
 
-  private URIListener = (_: Error | null, payload: { params: string[] }): void => {
-    this.events.emit(URI_AVAILABLE, payload.params[0])
+  private URIListener = (uri: string): void => {
+    this.events.emit(URI_AVAILABLE, uri)
   }
 
-  private async isomorphicInitialize(chainId = this.defaultChainId): Promise<void> {
-    if (this.eagerConnection) return
-
-    // because we can only use 1 url per chainId, we need to decide between multiple, where necessary
-    const rpc = Promise.all(
-      Object.keys(this.rpc).map(
-        async (chainId): Promise<[number, string]> => [
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
-      )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+  private async isomorphicInitialize(desiredChainId: number | undefined = this.defaultChainId): Promise<MockWalletConnectProvider> {
+    if (this.eagerConnection) return this.eagerConnection
+
+    if (this.provider) {
+      await this.deactivate()
```

Comment:
Why do we need to deactivate before initializing? We were doing this as part of activate before - should it still be in activate?

Reply:
So, this is a very special case that I think we should cover under unit tests. I tried to simplify the accessing logic and that's why I made this function return a Promise instead of being void.

That means... all the call sites that do `await this.isomorphicInitialize()` will now get a pointer to `eagerConnection` as long as it exists (which can be either a resolved `provider` or a Promise to the one that is loading). So it is safe to always call `await this.isomorphicInitialize()` and you will get a reference to the same provider, no matter where you access it from. I find it quite useful, but it may be an overkill.

I actually think this is dead code now, it is not possible to reach this code path under any circumstances. `eagerConnection` will always be defined (otherwise, we didn't loaded the module and there's 0 chance provider can be defined w/o eagerConnection.

Reply:
Now, answering your question:
> We were doing this as part of activate before - should it still be in activate?

I actually think that it was a dead code path before too:

https://github.com/Uniswap/web3-react/blob/a5fe4379bcf1e5f5c512348176ba54d66417dc09/packages/walletconnect/src/index.ts#L147-L154

First `if` condition catches all the cases when we're already connected. And we're connected when someone already called `activate` or `connectEagerly` before. In other words, we're not connected, when none of these were called before.

https://github.com/Uniswap/web3-react/blob/a5fe4379bcf1e5f5c512348176ba54d66417dc09/packages/walletconnect/src/index.ts#L158-L161

That means `deactivating` (especially with comparison to `chainId`) isn't really necessary, because: a) `provider` doesn't exist at this point anyway and b) there's no `chainId` active too. So, in other words, that had no actual impact.